### PR TITLE
[macOS] Multiview compositor

### PR DESCRIPTION
**Pull request by dkwingsmt**
This PR makes the macOS `FlutterCompositor` able to handle multiple views by adding methods `AddView` and `RemoveView`.

These two methods must take place on the main queue (platform thread), which is indeed the case, but extra effort is made to assert so. This is important because, due to the constraint of the macOS system that requires rendering to take place on the platform thread (hence `FlutterThreadSynchronizer`), no locks are needed to avoid race condition between `Add/RemoveView` and presenting.

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I signed the [CLA].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by dkwingsmt** [comment]
Also cc @loic-sharma

File: shell/platform/darwin/macos/framework/Source/FlutterCompositor.mm [hunk 1]
```diff
@@ -103,18 +112,27 @@
...
-                 ViewPresenter& presenter = presenters_[view_id];
-                 presenter.PresentPlatformViews(view, *layers_copy, platform_view_controller_);
-               }];
+  [view.surfaceManager presentSurfaces:surfaces
+                                atTime:presentation_time
+                                notify:^{
+                                  // Accessing presenters_ here does not need a lock to avoid race
+                                  // condition with AddView and RemoveView, as long as the latter
+                                  // take place on the platform thread, because presenting must take
+                                  // place on the platform thread due to macOS's requirement.
```
Comment (loic-sharma) [inline shell/platform/darwin/macos/framework/Source/FlutterCompositor.mm:121]: Could this assert that this is the platform thread?

**Review by loic-sharma — COMMENTED** [review]
This looks good to me, but please get approval from someone familiar with this area

File: shell/platform/darwin/macos/framework/Source/FlutterEngine.mm [hunk 2]
```diff
@@ -851,10 +860,6 @@ - (FlutterViewController*)viewController {
 }
 
 - (FlutterCompositor*)createFlutterCompositor {
```
Comment (cbracken) [inline shell/platform/darwin/macos/framework/Source/FlutterEngine.mm:862]: Is this still the right name for the method? "Create" usually implies allocation of some sort. Here what we're doing is more like initialisation the existing field and returning a pointer to it.

File: shell/platform/darwin/macos/framework/Source/FlutterCompositor.h [hunk 3]
```diff
@@ -40,12 +40,29 @@ class FlutterCompositor {
...
   // It must not be null, and is typically FlutterViewEngineProvider.
+  //
+  // The `main_queue` must be the event queue given to `FlutterThreadSynchronizer`,
+  // which is where the views' `FlutterSurfaceManager` invokes the callback in
+  // `presentSurfaces:atTime:notify:`. Typically `dispatch_get_main_queue()`.
   FlutterCompositor(id<FlutterViewProvider> view_provider,
                     FlutterTimeConverter* time_converter,
-                    FlutterPlatformViewController* platform_views_controller);
+                    FlutterPlatformViewController* platform_views_controller,
+                    dispatch_queue_t main_queue);
```
Comment (cbracken) [inline shell/platform/darwin/macos/framework/Source/FlutterCompositor.h:50]: Is there ever a case where this isn't the main queue returned by `dispatch_get_main_queue()`? If not, I think we should just test `dispatch_get_current_queue() == dispatch_get_main_queue()` where necessary.

**Review by cbracken — COMMENTED** [review]
Mostly lgtm - main question is why we need to pass in the dispatch queue.

File: shell/platform/darwin/macos/framework/Source/FlutterCompositor.h (hunk 3 above)
Comment (dkwingsmt) [reply shell/platform/darwin/macos/framework/Source/FlutterCompositor.h:50]: I'm mostly thinking about testability, although there really aren't many tests for `FlutterCompositor`. However, the `ThreadSynchronizer` accepts the queue as a replaceable parameter (which was added when I wrote tests for it). I think it's a clearer design if the same queue is passed in for both classes explicitly in production code.

File: shell/platform/darwin/macos/framework/Source/FlutterEngine.mm (hunk 2 above)
Comment (dkwingsmt) [reply shell/platform/darwin/macos/framework/Source/FlutterEngine.mm:862]: I think it's still a reasonable name. For whoever uses `FlutterEngine`, this method creates a  `embedderAPI/FlutterCompositor` object and returns it. How the object is memory-managed is implementation detail.

File: shell/platform/darwin/macos/framework/Source/FlutterCompositor.h (hunk 3 above)
Comment (dkwingsmt) [reply shell/platform/darwin/macos/framework/Source/FlutterCompositor.h:50]: Ok, `ThreadSynchronizer` has two initializers, one of which uses the main queue implicitly. Still, its internal implementation allows a replaceable queue. I'm ok with any suggestion.

**Review by cbracken — APPROVED** [review]
LGTM stamp from a Japanese personal seal

**Pull request merged**